### PR TITLE
Refactor to use message flow

### PR DESCRIPTION
**Pull request by stefanv**

### CODE REVIEW

**Comment by stefanv** [comment]
Please wait before merging.  My gut feeling is that this should not have passed.

**Comment by stefanv** [comment]
@bnaul It looks like the latest version of xarray broke API?

**Comment by bnaul** [comment]
Yeah, that's what https://github.com/mltsp/mltsp/pull/137 was for... did
you rebase after that was merged?

**Comment by stefanv** [comment]
I thought I did, will try again.

**Comment by bnaul** [comment]
This failed once because of that weird transient RethinkDB failure, but I re-built and it looks to have a real error as well...

**Comment by stefanv** [comment]
Thanks, working on it.

**Comment by bnaul** [comment]
Could you also update the "Starting the web app" section of `install.md` for the new front-end steps?

File: mltsp/cfg.py [hunk 1]
```diff
@@ -7,36 +7,33 @@
...
 
 # Load configuration
 config_files = [
     os.path.expanduser('~/.config/mltsp/mltsp.yaml'),
-    os.path.join(os.path.dirname(__file__), '../mltsp.yaml')
     ]
 
+config_files.extend(glob.glob(
+    os.path.join(os.path.dirname(__file__), '../mltsp-*.yaml')))
+config_files.extend(glob.glob(
```
Comment (bnaul) [inline mltsp/cfg.py]: Leftover?

File: mltsp/custom_feature_tools.py [hunk 2]
```diff
@@ -138,39 +145,31 @@ def parse_for_req_prov_params(script_fpath):
 
 def call_custom_functions(features_already_known, all_required_params,
                           all_provided_params, fnames_req_prov_dict,
-                          script_fpath=None):
+                          script_fpath):
```
Comment (bnaul) [inline mltsp/custom_feature_tools.py]: Could we get a docstring for this? :grimacing:

File: mltsp/custom_feature_tools.py [hunk 3]
```diff
@@ -138,39 +145,31 @@ def parse_for_req_prov_params(script_fpath):
...
 
-    all_required_params_copy = [x for x in all_required_params
-                                if x not in features_already_known]
-    for reqd_param in all_required_params_copy:
-        if reqd_param not in all_provided_params:
-            raise Exception((
-                "Not all of the required parameters are provided by the "
-                "functions in this script (required parameter '%s').") %
-                str(reqd_param))
+    all_required_params = [x for x in all_required_params
```
Comment (bnaul) [inline mltsp/custom_feature_tools.py]: `set(all_provided_params + features_already_known) >= set(all_required_params)` ?

File: mltsp/custom_feature_tools.py [hunk 4]
```diff
@@ -183,20 +182,19 @@ def call_custom_functions(features_already_known, all_required_params,
             reqs_provs_dict = fnames_req_prov_dict[funcname]
             reqs = reqs_provs_dict['requires']
             provs = reqs_provs_dict['provides']
-            if len(set(all_required_params_copy) & set(reqs)) > 0:
+            if len(set(all_required_params) & set(reqs)) > 0:
```
Comment (bnaul) [inline mltsp/custom_feature_tools.py]: `if (set(all_required_params) & set(reqs)):`

File: mltsp/custom_feature_tools.py [hunk 5]
```diff
@@ -183,20 +182,19 @@ def call_custom_functions(features_already_known, all_required_params,
...
                 func_queue.append(funcname)
             else:
                 func_rounds[str(i)].append(funcname)
-                all_required_params_copy = [x for x in all_required_params_copy
-                                            if x not in provs]
+                all_required_params = [x for x in all_required_params
```
Comment (bnaul) [inline mltsp/custom_feature_tools.py]: Set logic?

File: mltsp/custom_feature_tools.py [hunk 6]
```diff
@@ -416,25 +366,27 @@ def docker_extract_features(script_fpath, features_already_known):
...
 
     """
-    container_name = generate_random_str()
-    path_to_tmp_dir = make_tmp_dir()
+    container_id = str(uuid.uuid4())[:10]
 
-    copy_data_to_tmp_dir(path_to_tmp_dir, script_fpath, features_already_known)
+    tmp_dir = tempfile.mkdtemp(prefix='mltsp')
 
     try:
```
Comment (bnaul) [inline mltsp/custom_feature_tools.py]: What happens if this fails? Do we just keep going silently?

File: mltsp/custom_feature_tools.py [hunk 7]
```diff
@@ -466,19 +418,19 @@ def verify_new_script(script_fpath, docker_container=False):
...
 
     all_extracted_features = {}
-    no_docker = (os.getenv("MLTSP_NO_DOCKER") == "1")
-    if util.docker_images_available() and not no_docker:
-        print("Extracting features inside docker container...")
-        all_extracted_features = docker_extract_features(
-            script_fpath=script_fpath,
-            features_already_known=features_already_known)
-    elif no_docker:
+    if config['testing']['no_docker']:
```
Comment (bnaul) [inline mltsp/custom_feature_tools.py]: Is the `docker_container` variable here used at all?

File: mltsp/custom_feature_tools.py [hunk 8]
```diff
@@ -466,19 +418,19 @@ def verify_new_script(script_fpath, docker_container=False):
...
         print("WARNING - generating custom features WITHOUT docker container...")
         all_extracted_features = execute_functions_in_order(
             features_already_known=features_already_known,
             script_fpath=script_fpath)
-    elif not util.docker_images_available():
-        raise Exception("Docker image not available.")
+    elif util.docker_images_available():
```
Comment (bnaul) [inline mltsp/custom_feature_tools.py]: This looks like the same logic as `generate_custom_features`; could these be combined?

File: mltsp/featurize.py [hunk 9]
```diff
@@ -237,27 +238,31 @@ def featurize_time_series(times, values, errors=None, features_to_use=[],
...
         else:
             for t in times:
                 if isinstance(t, np.ndarray) and (t.ndim == 1 or 1 in t.shape):
-                    t[:] = np.linspace(0., cfg.DEFAULT_MAX_TIME, t.size)
+                    t[:] = np.linspace(0., config['mltsp']['DEFAULT_MAX_TIME'],
+                                       t.size)
                 else:
                     for t_i in t:
-                        t_i[:] = np.linspace(0., cfg.DEFAULT_MAX_TIME, t_i.size)
+                        t_i[:] = np.linspace(
```
Comment (bnaul) [inline mltsp/featurize.py]: Extra line break?

**Pull request merged**